### PR TITLE
Fix #611: Add line spacing to dxDrawText

### LINKED ISSUE #611: dxDrawText changing line spacing
**Is your feature request related to a problem? Please describe.**
Some fonts have wrong line spacing and would be nice if we could adjust it.

**Describe the solution you'd like**
Add option to change distance between lines in dxDrawText

**Describe alternatives you've considered**
-

**Additional context**
image
example of font that have to big line spacing

### PR DESCRIPTION
**Pull request by Allerek**
This PR will add lineHeight as last argument for dxDrawText, b.w.c. with older versions as 'false' 'nil' or '0' will use old system for line spacing.
This only works for colorcoded dxDrawText, as non-colorcoded uses different function for text rendering.
If someone manages to add this possilibites to non-colorcoded text, that would be great.
image

Test resource:
[drawTest.zip](https://github.com/multitheftauto/mtasa-blue/files/6387064/drawTest.zip)

Fixes https://github.com/multitheftauto/mtasa-blue/issues/611

### CODE REVIEW

File: Client/core/Graphics/CGraphics.cpp [hunk 1]
```diff
@@ -1138,7 +1138,8 @@ void CGraphics::DrawStringQueued(float fLeft, float fTop, float fRight, float fB
             fRight = floor(fRight);
             fBottom = floor(fBottom);
         }
-
+        if (fLineHeight == 0.0f)
```
Comment (tederis) [inline Client/core/Graphics/CGraphics.cpp]: It's better to use `std::optional` in here I think

Comment (tederis) [reply Client/core/Graphics/CGraphics.cpp]: Okay, I see some of modules(game_sa or multiplayer_sa) do not support C++17 features. You can't use std::optional here.

**Comment by Pirulax** [comment]
Yeah, apparently they've changed something in MSVC which caused naked functions to break.

File: Client/core/Graphics/CGraphics.cpp (hunk 1 above)
Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp]: Yeah, apparently they've changed something in MSVC C++17 which caused naked functions to break.

**Comment by Allerek** [comment]
> Yeah, apparently they've changed something in MSVC which caused naked functions to break.

Yeah, i've reverted back to float comparison method, just like it was meant in first commit, doing tests and its ready for review.

File: Client/core/Graphics/CGraphics.cpp (hunk 1 above)
Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp]: Btw I think he could use negative values to indicate something is wrong (as you obviously can't have negative spacing).

**Comment by Allerek** [comment]
image
Works just fine.

**Comment by Pirulax** [comment]
Definitely a nice feature, but imho people rarely do text break by themselves. I personally very rarely used any newlines by hand in dxDrawText.
One day hopefully we'll just replace DxDrawTextW with our own stuff, that way both this spacing stuff and colorcoded wordbreak would work.

**Comment by Allerek** [comment]
I think we should merge it for now, and then think about new drawing functions.

**Comment by theSarrum** [comment]
Personally, I think this is a bad practice. People will start using the `colorCoded` argument to be able to change the line spacing, but this will have a great impact on performance (more CPU load).

**Comment by Allerek** [comment]
> Personally, I think this is a bad practice. People will start using the `colorCoded` argument to be able to change the line spacing, but this will have a great impact on performance (more CPU load).

Thats their decision, we just allow them to change the spacing.

File: Client/core/Graphics/CGraphics.cpp (hunk 1 above)
Comment (Allerek) [reply Client/core/Graphics/CGraphics.cpp]: > Btw I think he could use negative values to indicate something is wrong (as you obviously can't have negative spacing).

Well, you can(last example in screenshot below, where the last line became the first)

**Comment by patrikjuvonen** [comment]
Thanks for your work with this!

Unfortunately you seem to have deleted the original repository you created this pull request with.

Please move your changes under a feature branch in your repository and recreate this pull request so we can test it properly.

**Pull request closed without merging**